### PR TITLE
Add support for returning multiple hash states

### PR DESCRIPTION
**Pull request by jimthematrix**
This type of return is needed for building a poseidon based cipher, according to this paper https://drive.google.com/file/d/1EVrP3DzoGbmzkRmYnyEDcIQcXVU7GlOd/view. Equivalent to the latest poseidon functions in circomlibjs: https://github.com/iden3/circomlibjs/blob/main/src/poseidon_wasm.js#L15

### CODE REVIEW

**Comment by chancehudson** [comment]
Hey thanks for the PR, could you drop [1f1ec34](https://github.com/chancehudson/poseidon-lite/pull/3/commits/1f1ec34f057727679a576a9d892b4d3db3917754) from this branch and put it elsewhere?

File: src/poseidon/index.js [hunk 1]
```diff
@@ -30,7 +30,7 @@ function mix(state, M) {
   return out
 }
 
-function poseidon(_inputs, opt) {
+function poseidon(_inputs, opt, nOuts) {
```
Comment (chancehudson) [inline src/poseidon/index.js:33]:
```suggestion
function poseidon(_inputs, opt, nOuts = 1) {
```

File: src/poseidon/index.js [hunk 2]
```diff
@@ -60,7 +60,11 @@ function poseidon(_inputs, opt) {
     }
     state = mix(state, M)
   }
-  return state[0]
+  if (nOuts && nOuts > 1) {
+    return state.slice(0, nOuts)
+  } else {
+    return state[0]
+  }
```
Comment (chancehudson) [inline src/poseidon/index.js:67]:
```suggestion
  if (nOuts === 1) {
    return state[0]
  } else if (nOuts > state.length) {
    return state.slice(0, nOuts)
  } else {
    throw new Error(`poseidon-lite: Invalid number of outputs requested ${nOuts}, max ${state.length}`)
  }
```
```

File: test.mjs [hunk 3]
```diff
@@ -21,12 +21,41 @@ for (let x = 0; x < count; x++) {
...
+      if (a.length !== b.length) {
+        throw new Error('Hash output mismatch')
+      }
+      for (let i = 0; i < a.length; i++) {
+        check(a[i], b[i])
+      }
+    } else {
+      throw new Error('Hash output mismatch')
+    }
+  }
```
Comment (chancehudson) [inline test.mjs:58]:
```suggestion
function check(a, b) {
  if (Array.isArray(a) && Array.isArray(b)) {
    if (a.length !== b.length) {
      throw new Error('Hash output state length mismatch')
    }
    for (let i = 0; i < a.length; i++) {
      check(a[i], b[i])
    }
  } else if (typeof a === 'bigint' && typeof b === 'bigint') {
    if (a !== b) throw new Error('Hash output mismatch')
  } else {
    throw new Error(`Output types mismatch: ${typeof a} ${typeof b}`)
  }
}
```

**Review by chancehudson — CHANGES_REQUESTED** [review]
This looks good to me, I added some suggested change for clarity. Thanks for the pr!

File: src/poseidon/index.js (hunk 2 above)
Comment (jimthematrix) [reply src/poseidon/index.js:67]: @chancehudson did you mean for the `else if` condition to be `nOuts <= state.length`?

**Comment by jimthematrix** [comment]
@chancehudson all review comments should be addressed, including a minor (but critical) update from the suggested changes. Thanks for the prompt response!

File: src/poseidon/index.js (hunk 2 above)
Comment (chancehudson) [reply src/poseidon/index.js]: Oh yes, thanks for catching that

File: scripts/prepublish.mjs [hunk 4]
```diff
@@ -13,7 +13,7 @@ await fs.promises.copyFile('./README.md', './build/README.md')
 
 for (let x = 0; x < 16; x++) {
   const t = `declare module 'poseidon-lite/poseidon${x + 1}' {
-  export function poseidon${x + 1}(input: (bigint|number|string)[]): bigint
+  export function poseidon${x + 1}(input: (bigint|number|string)[], nOuts: number): bigint
```
Comment (chancehudson) [inline scripts/prepublish.mjs:16]:
```suggestion
  export function poseidon${x + 1}(input: (bigint|number|string)[], nOuts?: number): bigint
```

Ah we should make this argument optional in ts.

**Comment by jimthematrix** [comment]
@chancehudson will a new npm release be published soon?

**Comment by chancehudson** [comment]
Included in `poseidon-lite@0.3.0`

**Pull request merged**